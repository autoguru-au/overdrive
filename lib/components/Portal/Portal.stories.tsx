import { Meta, StoryObj } from '@storybook/react';
import React from 'react';

import { Heading } from '../Heading';
import { Stack } from '../Stack';
import { Text } from '../Text';

import { Portal } from '.';

export default {
	title: 'Utility/Portal',
	component: Portal,
	parameters: {
		chromatic: { disable: true },
	},
	argTypes: {
		children: {
			control: { disable: true },
		},
		container: {
			control: { disable: true },
		},
	},
} satisfies Meta<typeof Portal>;

type Story = StoryObj<typeof Portal>;

const containerEl = document?.getElementsByTagName('body')[0];

export const Standard: Story = {
	args: {
		container: containerEl,
		children: <Text colour="primary">Im in a portal at the root.</Text>,
	},
};

export const Nested: Story = {
	args: {
		container: containerEl,
		children: (
			<Stack space="5">
				<Heading is="h3" colour="information">
					test child 1
				</Heading>
				<Portal container={containerEl}>
					<Text colour="primary">test child 2</Text>
				</Portal>
			</Stack>
		),
	},
};
