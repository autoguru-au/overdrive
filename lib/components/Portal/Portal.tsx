import React, { forwardRef, useEffect, useLayoutEffect, useState } from 'react';
import { createPortal } from 'react-dom';

import { isBrowser, isHtmlElement, setRef } from '../../utils';
import { useTheme } from '../OverdriveProvider';

export interface PortalProps {
	children?: React.ReactNode;
	container?: HTMLElement;
	noThemedWrapper?: boolean;
}

type RefValue<T> = T extends React.RefObject<infer T> ? T : never;

function Portal(
	{ children, container, noThemedWrapper }: PortalProps,
	ref: React.Ref<typeof container>,
) {
	const { themeClass, portalMountPoint } = useTheme();
	const [mountPoint, setMountPoint] = useState<HTMLElement | null>(null);
	const [mountNode, setMountNode] = useState<RefValue<typeof ref> | null>(
		null,
	);

	// use local state to set the mount point if it populates after the component mounts
	useEffect(() => {
		if (!isBrowser || !portalMountPoint?.current) return;

		if (mountPoint !== portalMountPoint.current) {
			setMountPoint(portalMountPoint.current);
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [portalMountPoint?.current]);

	useLayoutEffect(() => {
		if (!isBrowser) return;
		if (isHtmlElement(mountPoint)) {
			setMountNode(mountPoint);
			return;
		}
		setMountNode(isHtmlElement(container) ? container : document.body);
	}, [container, mountPoint]);

	useLayoutEffect(() => {
		if (!isBrowser) return;

		if (mountNode) {
			setRef(ref, mountNode);
		}

		return () => {
			setRef(ref, null);
		};
	}, [ref, mountNode]);

	if (!mountNode) return null;

	return noThemedWrapper
		? createPortal(children, mountNode)
		: createPortal(<div className={themeClass}>{children}</div>, mountNode);
}

const _Portal = forwardRef(Portal);
export { _Portal as Portal };

export default Portal;
