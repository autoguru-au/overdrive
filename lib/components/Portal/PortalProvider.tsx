import React, {
	Context,
	createContext,
	FunctionComponent,
	useContext,
} from 'react';

interface Props {
	children;
}
export const PortalContext: Context<{
	portalInstance: Element;
}> = createContext(null);

export const usePortalContext = () => useContext(PortalContext);

const buildPortalElement = (): Element => {
	const element: Element = document.createElement('div');
	element.setAttribute('id', `od-portal-${Math.round(Math.random() * 1e6)}`);
	document.body.append(element);

	return element;
};

export const PortalProvider: FunctionComponent<Props> = ({ children }) => {
	const portalContext = usePortalContext();
	const context = portalContext || {
		portalInstance: buildPortalElement(),
	};

	return (
		<PortalContext.Provider value={context}>
			{children}
		</PortalContext.Provider>
	);
};
