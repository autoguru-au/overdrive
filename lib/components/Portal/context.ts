import { Context, createContext } from 'react';

export const PortalContext: Context<{
	portalInstance: Element;
}> = createContext(null);
