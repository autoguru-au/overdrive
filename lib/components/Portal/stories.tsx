import { ComponentMeta, ComponentStory } from '@storybook/react';
import * as React from 'react';
import { ComponentProps } from 'react';

import { Heading } from '../Heading';
import { Stack } from '../Stack';
import { Text } from '../Text';

import { Portal } from '.';

export default {
	title: 'Utility/Portal',
	component: Portal,
	parameters: {
		chromatic: { disable: true },
	},
	argTypes: {
		children: {
			control: {
				disable: true,
			},
		},
		container: {
			control: {
				disable: true,
			},
		},
	},
} as ComponentMeta<typeof Portal>;

const StandardTemplate: ComponentStory<typeof Portal> = (...args) => (
	<Portal {...args}>
		<Text colour="primary">Im in a portal at the root.</Text>
	</Portal>
);

const NestedTemplate: ComponentStory<typeof Portal> = (...args) => (
	<Portal {...args}>
		<Stack space="5">
			<Heading is="h3" colour="information">
				test child 1
			</Heading>
			<Portal {...args}>
				<Text colour="primary">test child 2</Text>
			</Portal>
		</Stack>
	</Portal>
);

const standardProps: Omit<ComponentProps<typeof Portal>, 'children'> = {
	container: document?.getElementsByTagName('body')[0],
};
export const standard = StandardTemplate.bind(standardProps);
standard.args = standardProps;

export const nested = NestedTemplate.bind(standardProps);
nested.args = standardProps;
