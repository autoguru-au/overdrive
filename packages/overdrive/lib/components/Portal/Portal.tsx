import type { ReactNode, Ref, RefObject } from 'react';
import { forwardRef, useLayoutEffect, useState } from 'react';
import { createPortal } from 'react-dom';

import { isHtmlElement, setRef } from '../../utils';

export interface Props {
	children?: ReactNode;
	container?: Element;
}

type RefValue<T> = T extends RefObject<infer T> ? T : never;

function Portal({ children, container }: Props, ref: Ref<typeof container>) {
	const [mountNode, setMountNode] = useState<RefValue<typeof ref> | null>(
		null,
	);

	useLayoutEffect(() => {
		setMountNode(isHtmlElement(container) ? container : document.body);
	}, [container]);

	useLayoutEffect(() => {
		if (mountNode) {
			setRef(ref, mountNode);
		}

		return () => void setRef(ref, null);
	}, [ref, mountNode]);

	return mountNode ? createPortal(children, mountNode) : null;
}

const _Portal = forwardRef(Portal);
export { _Portal as Portal };
