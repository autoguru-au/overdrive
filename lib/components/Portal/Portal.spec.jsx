import * as React from 'react';
import { render } from '@testing-library/react';
import { Portal } from './Portal';

describe('<Portal />', () => {
	it('should not throw', () => {
		expect(() => render(<Portal />)).not.toThrow();
	});

	it('should match the snapshot', () => {
		expect(render(<Portal />).container.firstChild).toMatchSnapshot();
	});
});
