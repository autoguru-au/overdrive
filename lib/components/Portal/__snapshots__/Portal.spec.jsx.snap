// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portal /> should match the snapshot 1`] = `null`;
